dbms/src/Storages/StorageFile.cpp

```diff
@@ -0,0 +1,232 @@
+		}
+		else // Is DB's file
+		{
+			path = getTablePath(db_dir_path, table_name, format_name);
+			is_db_table = true;
+			Poco::File(Poco::Path(path).parent()).createDirectories();
+		}
+	}
+	else
+	{
+		is_db_table = false;
+		use_table_fd = true;
+		table_fd_init_offset = lseek(table_fd, 0, SEEK_CUR);
+	}
+}
+
+
+class StorageFileBlockInputStream : public IProfilingBlockInputStream
+{
+public:
+
+	StorageFileBlockInputStream(StorageFile & storage_, const Context & context, size_t max_block_size)
+	: storage(storage_), lock(storage.rwlock, storage.use_table_fd)
+	{
+		if (storage.use_table_fd)
+		{
+			/// We could use common ReadBuffer and WriteBuffer in storage to leverage cache
+			///  and add ability to seek unseekable files, but cache sync isn't supported.
+
+			if (storage.table_fd_was_used) /// We need seek to initial position
+			{
+				if (storage.table_fd_init_offset < 0)
+					throw Exception("File descriptor isn't seekable, inside " + storage.getName(), ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
+
+				/// ReadBuffer's seek() doesn't make sence, since cache is empty
+				if (lseek(storage.table_fd, storage.table_fd_init_offset, SEEK_SET) < 0)
+					throw Exception("Cannot seek file descriptor, inside " + storage.getName(), ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
+			}
+
+			storage.table_fd_was_used = true;
+			read_buf = std::make_unique<ReadBufferFromFileDescriptor>(storage.table_fd);
+		}
+		else
+		{
+			read_buf = std::make_unique<ReadBufferFromFile>(storage.path);
+		}
+
+		input_formatter = FormatFactory().getInput(storage.format_name, *read_buf, storage.getSampleBlock(), context, max_block_size);
+	}
+
+	String getName() const override
+	{
+		return storage.getName();
+	}
+
+	String getID() const override
+	{
+		std::stringstream res_stream;
+		res_stream << "File(" << &storage << ")";
+		return res_stream.str();
+	}
+
+	Block readImpl() override
+	{
+		return input_formatter->read();
+	}
+
+	void readPrefixImpl() override
+	{
+		input_formatter->readPrefix();
+	}
+
+	void readSuffixImpl() override
+	{
+		input_formatter->readSuffix();
+	}
+
+private:
+	StorageFile & storage;
+	Poco::ScopedRWLock lock;
+	Block sample_block;
+	std::unique_ptr<ReadBufferFromFileDescriptor> read_buf;
+	BlockInputStreamPtr input_formatter;
+};
+
+
+BlockInputStreams StorageFile::read(
+	const Names & column_names,
+	ASTPtr query,
+	const Context & context,
+	const Settings & settings,
+	QueryProcessingStage::Enum & processed_stage,
+	size_t max_block_size,
+	unsigned threads)
+{
+	return BlockInputStreams(1, std::make_shared<StorageFileBlockInputStream>(*this, context, max_block_size));
+}
+
+
+class StorageFileBlockOutputStream : public IBlockOutputStream
+{
+public:
+
+	StorageFileBlockOutputStream(StorageFile & storage_)
+	: storage(storage_), lock(storage.rwlock)
+	{
+		if (storage.use_table_fd)
+		{
+			/// TODO: more detail checks: 1) fd is writeable 2) points to the end of data (if possible)
+			if (storage.table_fd_was_used)
```

Comment:
I worried about misleading in such case:
1) select -> data1; insert data2; select -> data1 + data2;
3) insert data2 (overwrites data1); select -> data2;

But such cases is possible when FD is binded to real file, which is quite rare.

So, multiple writes and write after read for FD should be possible, so there are no any checks are required.